rfcs/0005-collection-context.md

```diff
@@ -0,0 +1,494 @@
+For example:
+
+```rust
+struct Data1 {
+    other: Data2
+}
+
+struct Data2 {
+    other: Data1
+}
+```
+
+In these instances, the compiler performs an [occurs check](https://papl.cs.brown.edu/2019/Type_Inference.html)
+to check whether the same type variable occurs on both sides and, if it does, it produces a compile error.
+
+I propose that we use an occurs check detect to recursive data structures until we fully support an equivalent
+to something like the Rust `Box` type, which would eliminate this problem.
+
+## Program Visualization
+
+The `CollectionContext` uses [petgraph](https://crates.io/crates/petgraph) under the hood, meaning we get
+to benefit from how awesome this library is. So, one of the cool things we can do is we can actually print
+the `CollectionContext` to a `.dot` file and visualize the graph.
+
+Given this simple generic function example (`UNK` is the unknown type in the prototype):
+
+```rust
+fn F<T>(n: T) -> T {
+  let x: T = n;
+  let y: UNK = 5u8;
+  return x;
+};
+
+fn main() -> () {
+  let foo: UNK = F(1u32);
+  let bar: UNK = F(1u64);
+};
+```
+
+We can visualize the program before type inference:
+
+![oopsies image not showing](../files/0005-before.svg)
+
+and after type inference, notice that the `UNK`'s have disappeared:
+
+![oopsies image not showing](../files/0005-after.svg)
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+I would say developer time and lots of merge conflicts. At the very least, disentangling AST transformation
+from type inference and type checking is a must-do, in my opinion, even if the collection context
+does not get accepted in this form.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+# Prior art
+
+[prior-art]: #prior-art
+
+Rust separates the ideas of
+[type collection](https://rustc-dev-guide.rust-lang.org/type-checking.html#type-collection),
+[type checking](https://rustc-dev-guide.rust-lang.org/type-checking.html#type-checking),
+and [type inference](https://rustc-dev-guide.rust-lang.org/type-inference.html). And these are
+all [disentangled from 'AST lowering'](https://rustc-dev-guide.rust-lang.org/overview.html#hir-lowering).
+
+Relating to the `CollectionContext`, the Rust compiler uses
+[queries for demand-driven development](https://rustc-dev-guide.rust-lang.org/query.html). Admittedly,
+I'm not as familiar with this on a technical level as I would like to be, but my guess is that they are
+using a graph-based approach (combined with the Rust `DefId`) and to achieve the memoization and caching.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+I'd really like to get feedback on how the `CollectionContext` could be improved for use with the LSP.
```

Comment:
I can't really think of anything that could be improved at this stage. It all seems relatively straightforward.

I can see a situation in LSP where we have something like:

```
fn foo(bob: Bob) {
    let x = bob.
}
```

Could there be a way for the LSP to query the CC with the unfinished `bob` token to look up the closest edge in the graph? This way we could then go to that type and look up it's methods for autocompletion etc..